From a11fc69dd21a9b91089506225a182432b7c4cd84 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Tue, 5 Nov 2024 09:58:33 +0100
Subject: [PATCH 1343/1695] drivers: stm32_risab: reorder CID filtering
 configuration application

When applying RISAB CID filtering configuration, start by applying
read/write/priv by CID rights. This fixes an issue where a data cache
eviction occurs after the CID filtering is enabled but before
read/write/priv rights are set.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 1ba1ec6d72a5e3d9919ba675d58001a02407c39b drivers: stm32_risab: reorder CID filtering configuration application]
Change-Id: I43eed79aa7d996c3ac20ca01c95a7f33fce58dd8
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/416408
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/firewall/stm32_risab.c | 169 +++++++++++-----------------
 1 file changed, 67 insertions(+), 102 deletions(-)

diff --git a/core/drivers/firewall/stm32_risab.c b/core/drivers/firewall/stm32_risab.c
index 83a4aab2f..ffc0bdaa3 100644
--- a/core/drivers/firewall/stm32_risab.c
+++ b/core/drivers/firewall/stm32_risab.c
@@ -165,11 +165,18 @@ static bool regs_access_granted(struct stm32_risab_pdata *risab_d,
 	uint32_t cidcfgr = io_read32(risab_base(risab_d) +
 				     _RISAB_PGy_CIDCFGR(first_page));
 
+	if (virt_to_phys((void *)risab_base(risab_d)) == RISAB1_BASE ||
+	    virt_to_phys((void *)risab_base(risab_d)) == RISAB2_BASE)
+		return true;
+
+	/* No CID filtering */
+	if (!(cidcfgr & _RISAB_PG_CIDCFGR_CFEN))
+		return true;
+
 	/* Trusted CID access */
 	if (is_tdcid &&
-	    ((cidcfgr & _RISAB_PG_CIDCFGR_CFEN &&
-	      !(cidcfgr & _RISAB_PG_CIDCFGR_DCEN)) ||
-	     !(cidcfgr & _RISAB_PG_CIDCFGR_CFEN)))
+	    (cidcfgr & _RISAB_PG_CIDCFGR_CFEN &&
+	     !(cidcfgr & _RISAB_PG_CIDCFGR_DCEN)))
 		return true;
 
 	/* Delegated CID access check */
@@ -289,6 +296,51 @@ static void set_cid_priv_conf(struct stm32_risab_pdata *risab_d,
 	}
 }
 
+static TEE_Result set_rif_registers(struct stm32_risab_pdata *risab,
+				    unsigned int reg_idx)
+{
+	struct stm32_risab_rif_conf *subr_cfg = NULL;
+
+	assert(&risab->subr_cfg[reg_idx]);
+
+	subr_cfg = &risab->subr_cfg[reg_idx];
+
+	/*
+	 * This sequence will generate an IAC if the CID filtering
+	 * configuration is inconsistent with these desired rights
+	 * to apply.
+	 */
+	if (!regs_access_granted(risab, reg_idx))
+		return TEE_ERROR_ACCESS_DENIED;
+
+	set_block_dprivcfgr(risab, subr_cfg);
+	set_block_seccfgr(risab, subr_cfg);
+
+	/*
+	 * Grant page access to some CIDs, in read and/or write, and the
+	 * necessary privilege level.
+	 */
+	set_read_conf(risab, subr_cfg);
+	set_write_conf(risab, subr_cfg);
+	set_cid_priv_conf(risab, subr_cfg);
+
+	if (virt_to_phys((void *)risab_base(risab)) != RISAB1_BASE &&
+	    virt_to_phys((void *)risab_base(risab)) != RISAB2_BASE) {
+		/* Delegate RIF configuration or not */
+		if (!is_tdcid)
+			DMSG("Cannot set %s CID config for region %u",
+			     risab->risab_name, reg_idx);
+		else
+			set_cidcfgr(risab, subr_cfg);
+	} else {
+		set_cidcfgr(risab, subr_cfg);
+	}
+
+	dsb();
+
+	return TEE_SUCCESS;
+}
+
 static void apply_rif_config(struct stm32_risab_pdata *risab_d)
 {
 	vaddr_t base = risab_base(risab_d);
@@ -315,37 +367,8 @@ static void apply_rif_config(struct stm32_risab_pdata *risab_d)
 	}
 
 	for (i = 0; i < risab_d->nb_regions_cfged; i++) {
-		set_block_dprivcfgr(risab_d, &risab_d->subr_cfg[i]);
-
-		if (virt_to_phys((void *)base) != RISAB1_BASE &&
-		    virt_to_phys((void *)base) != RISAB2_BASE) {
-			/* Delegate RIF configuration or not */
-			if (!is_tdcid)
-				DMSG("Cannot set %s CID config for region %u",
-				     risab_d->risab_name, i);
-			else
-				set_cidcfgr(risab_d, &risab_d->subr_cfg[i]);
-
-			if (!regs_access_granted(risab_d, i))
-				panic();
-		} else {
-			set_cidcfgr(risab_d, &risab_d->subr_cfg[i]);
-		}
-
-		/*
-		 * This sequence will generate an IAC if the CID filtering
-		 * configuration is inconsistent with these desired rights
-		 * to apply. Start by default setting security configuration
-		 * for all blocks.
-		 */
-		set_block_seccfgr(risab_d, &risab_d->subr_cfg[i]);
-
-		/* Grant page access to some CIDs, in read and/or write */
-		set_read_conf(risab_d, &risab_d->subr_cfg[i]);
-		set_write_conf(risab_d, &risab_d->subr_cfg[i]);
-
-		/* For each granted CID define privilege access per page */
-		set_cid_priv_conf(risab_d, &risab_d->subr_cfg[i]);
+		if (set_rif_registers(risab_d, i))
+			panic();
 	}
 
 	clk_disable(risab_d->clock);
@@ -707,14 +730,10 @@ out:
 static TEE_Result stm32_risab_reconfigure_region(struct firewall_query *fw,
 						 paddr_t paddr, size_t size)
 {
-	struct stm32_risab_rif_conf reg_conf = { };
 	struct stm32_risab_pdata *risab = NULL;
 	TEE_Result res = TEE_ERROR_GENERIC;
 	paddr_t sub_region_offset = 0;
-	unsigned int first_page = 0;
-	uint32_t old_cidcfgr = 0;
 	uint32_t exceptions = 0;
-	uint32_t old_dpriv = 0;
 	unsigned int i = 0;
 
 	assert(fw->ctrl->priv);
@@ -754,70 +773,35 @@ static TEE_Result stm32_risab_reconfigure_region(struct firewall_query *fw,
 		parse_risab_rif_conf(risab, &risab->subr_cfg[i], fw->args[0],
 				     false /*!check_overlap*/);
 
-		reg_conf = risab->subr_cfg[i];
 		break;
 	}
 
 	if (i == risab->nb_regions_cfged)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	first_page = risab->subr_cfg[i].first_page;
 	res = clk_enable(risab->clock);
 	if (res)
 		return res;
 
 	DMSG("Reconfiguring %s pages %u-%u: %s, %s, CID attributes: %#"PRIx32", R:%#"PRIx32", W:%#"PRIx32", P:%#"PRIx32"",
-	     risab->risab_name, reg_conf.first_page, reg_conf.first_page +
-	     reg_conf.nb_pages_cfged - 1,
-	     reg_conf.seccfgr ? "Secure" : "Non secure",
-	     reg_conf.dprivcfgr ? "Default priv" : "Default unpriv",
-	     reg_conf.cidcfgr,
+	     risab->risab_name, risab->subr_cfg[i].first_page,
+	     risab->subr_cfg[i].first_page +
+	     risab->subr_cfg[i].nb_pages_cfged - 1,
+	     risab->subr_cfg[i].seccfgr ? "Secure" : "Non secure",
+	     risab->subr_cfg[i].dprivcfgr ? "Default priv" : "Default unpriv",
+	     risab->subr_cfg[i].cidcfgr,
 	     fw->args[0] & RISAB_RLIST_MASK,
 	     fw->args[0] & RISAB_WLIST_MASK,
 	     fw->args[0] & RISAB_PLIST_MASK);
 
 	exceptions = cpu_spin_lock_xsave(&risab->conf_lock);
 
-	old_dpriv = io_read32(risab_base(risab) +
-			      _RISAB_PGy_PRIVCFGR(first_page));
-	set_block_dprivcfgr(risab, &reg_conf);
-
-	/* Delegate RIF configuration or not */
-	if (!is_tdcid) {
-		DMSG("Cannot set %s CID configuration for region %u",
-		     risab->risab_name, i);
-	} else {
-		old_cidcfgr = io_read32(risab_base(risab) +
-					_RISAB_PGy_CIDCFGR(first_page));
-		set_cidcfgr(risab, &reg_conf);
-	}
-
-	if (!regs_access_granted(risab, i)) {
-		/* Restore old configuration before returning error */
-		if (is_tdcid) {
-			reg_conf.cidcfgr = old_cidcfgr;
-			set_cidcfgr(risab, &reg_conf);
-		}
-		reg_conf.dprivcfgr = old_dpriv;
-		set_block_dprivcfgr(risab, &reg_conf);
+	if (set_rif_registers(risab, i)) {
 		cpu_spin_unlock_xrestore(&risab->conf_lock, exceptions);
 		clk_disable(risab->clock);
 		return TEE_ERROR_ACCESS_DENIED;
 	}
 
-	/*
-	 * This sequence will generate an IAC if the CID filtering configuration
-	 * is inconsistent with these desired rights to apply. Start by default
-	 * setting security configuration for all blocks.
-	 */
-	set_block_seccfgr(risab, &reg_conf);
-
-	/* Grant page access to some CIDs, in read and/or write */
-	set_read_conf(risab, &reg_conf);
-	set_write_conf(risab, &reg_conf);
-
-	/* For each granted CID define privilege access per page */
-	set_cid_priv_conf(risab, &reg_conf);
 	cpu_spin_unlock_xrestore(&risab->conf_lock, exceptions);
 
 	clk_disable(risab->clock);
@@ -827,7 +811,7 @@ static TEE_Result stm32_risab_reconfigure_region(struct firewall_query *fw,
 
 static TEE_Result stm32_risab_pm_resume(struct stm32_risab_pdata *risab)
 {
-	size_t i = 0;
+	unsigned int i = 0;
 
 	if (is_tdcid) {
 		if (risab->base.pa == RISAB6_BASE)
@@ -837,27 +821,8 @@ static TEE_Result stm32_risab_pm_resume(struct stm32_risab_pdata *risab)
 	}
 
 	for (i = 0; i < risab->nb_regions_cfged; i++) {
-		/* Restoring RISAB RIF configuration */
-		set_block_dprivcfgr(risab, &risab->subr_cfg[i]);
-
-		if (!is_tdcid)
-			DMSG("Cannot set %s CID configuration for region %zu",
-			     risab->risab_name, i);
-		else
-			set_cidcfgr(risab, &risab->subr_cfg[i]);
-
-		if (!regs_access_granted(risab, i))
-			continue;
-
-		/*
-		 * This sequence will generate an IAC if the CID filtering
-		 * configuration is inconsistent with these desired rights
-		 * to apply.
-		 */
-		set_block_seccfgr(risab, &risab->subr_cfg[i]);
-		set_read_conf(risab, &risab->subr_cfg[i]);
-		set_write_conf(risab, &risab->subr_cfg[i]);
-		set_cid_priv_conf(risab, &risab->subr_cfg[i]);
+		if (set_rif_registers(risab, i))
+			panic();
 	}
 
 	disable_srwiad_if_unset(risab);
-- 
2.47.2

